gluten-celeborn/velox/src/main/scala/org/apache/spark/shuffle/VeloxCelebornColumnarBatchSerializer.scala

```diff
@@ -96,128 +98,155 @@ private class CelebornColumnarBatchSerializerInstance(
+        }
+        gotNext = false
+        nextValue
       }
+    }
 
-      override def readKey[T: ClassTag](): T = {
-        // We skipped serialization of the key in writeKey(), so just return a dummy value since
-        // this is going to be discarded anyways.
-        null.asInstanceOf[T]
-      }
+    override def asIterator: Iterator[Any] = {
+      // This method is never called by shuffle code.
+      throw new UnsupportedOperationException
+    }
 
-      @throws(classOf[EOFException])
-      override def readValue[T: ClassTag](): T = {
-        if (cb != null) {
-          cb.close()
-          cb = null
-        }
-        val batch = {
-          val maybeBatch =
-            try {
-              wrappedOut.next()
-            } catch {
-              case ioe: IOException =>
-                this.close()
-                logError("Failed to load next RecordBatch", ioe)
-                throw ioe
-            }
-          if (maybeBatch == null) {
-            // EOF reached
-            this.close()
-            throw new EOFException
+    override def readKey[T: ClassTag](): T = {
+      // We skipped serialization of the key in writeKey(), so just return a dummy value since
+      // this is going to be discarded anyways.
+      null.asInstanceOf[T]
+    }
+
+    @throws(classOf[EOFException])
+    override def readValue[T: ClassTag](): T = {
+      if (cb != null) {
+        cb.close()
+        cb = null
+      }
+      val batch = {
+        val maybeBatch =
+          try {
+            wrappedOut.next()
+          } catch {
+            case ioe: IOException =>
+              this.close()
+              logError("Failed to load next RecordBatch", ioe)
+              throw ioe
           }
-          maybeBatch
+        if (maybeBatch == null) {
+          // EOF reached
+          this.close()
+          throw new EOFException
         }
-        val numRows = batch.numRows()
-        logDebug(s"Read ColumnarBatch of $numRows rows")
-        numBatchesTotal += 1
-        numRowsTotal += numRows
-        cb = batch
-        cb.asInstanceOf[T]
+        maybeBatch
+      }
+      val numRows = batch.numRows()
+      logDebug(s"Read ColumnarBatch of $numRows rows")
+      numBatchesTotal += 1
+      numRowsTotal += numRows
+      cb = batch
+      cb.asInstanceOf[T]
+    }
+
+    override def readObject[T: ClassTag](): T = {
+      // This method is never called by shuffle code.
+      throw new UnsupportedOperationException
+    }
+
+    override def close(): Unit = {
+      if (!closeCalled.compareAndSet(false, true)) {
+        return
       }
+      // Would remove the resource object from registry to lower GC pressure.
+      TaskResources.releaseResource(resourceId)
+    }
 
-      override def readObject[T: ClassTag](): T = {
-        // This method is never called by shuffle code.
-        throw new UnsupportedOperationException
+    override def release(): Unit = {
```

Comment:
What's the benefit of implementing `TaskResource` interface? The code looks a bit confusing, that `close()` calls `releaseResource()` while `release()` calls `close0()`.

Reply:
> The code looks a bit confusing, that close() calls releaseResource() while release() calls close0().

Yes this is a little bit implicit to me and be worth to be discussed.

Let's summarize the core change first. Previous code:

``` scala
override def close(): Unit = {
  if (!isClosed) {
    TaskResources.removeResource(resourceId)
    close0()
    isClosed = true
  }
}
```
and the current code:
```scala
override def close(): Unit = {
  if (!closeCalled.compareAndSet(false, true)) {
    return
  }
  // Would remove the resource object from registry to lower GC pressure.
  TaskResources.releaseResource(resourceId)
}
```

So basically in the patch I consolidated the two operations `removeResource` and `close0`, into a single API `releaseResource`. You can consider `removeResource` + `close0` a common pattern since `removeResource` omits running `release()` method of the registered TaskResource. So after calling `removeResource` you should call `close0` which does exactly the same with resource's `release()` method. Since the code here can be a common pattern, I added `releaseResource` for better encapsulation. Also, exposing `removeResouce` API may lead to unexpected usages, e.g. removing a resource and go around the resource clean-up procedure, we should only allow this kind of usage for certain strong reasons.